### DISCUSSION TITLE
Modify the resampling method in load

### DISCUSSION THREAD

**Discussion by bcerripromethee** [sertit/eoreader#103] [Q&A] [answered]
Hi, 

Is there a way to replace the resampling method, which is nearest, to bilinear (for example) in the load function ?

Actually when i use the function : 
`prod.load(band, window=window, resolution=10, resampling=Resampling.bilinear)`

By doing this, i'm getting this error :
`eoreader.utils.read() got multiple values for keyword argument 'resampling'`

My first idea is to reach out your function in your code and modify it, but it's not really "clean", and i would have to reproduce it at each update of eoreader.

Do you have an idea ? 
Thanks,
Bastien

**Comment by remi-braun** [comment]
Hi Bastien,

Are you sure the resampling method is set to nearest by default ?
This shouldn't be the case (see the `_read_band` method of the satellite you are using)

For the moment I am managing this by forcing a method (so I should document it) but the philosophy is: 
- for bands that should be continuous, loaded in float (like spectral bands) : using bilinear method
- for masks and other classified rasters : nearest

**Reply by bcerripromethee** [reply]
Oh I thought the method `collocate` was the one used for reading bands. Because collocates is using the nearest method.

**Reply by bcerripromethee** [reply]
Actually, i'm trying to understand why i do have different values over my Sentinel-2 SWIR1 band between a **.SAFE** product created by eoreader and a **Element 84** product. The values are close but not exactly the same, this is really weird and it affects spectral indices after.

When i read the SWIR1 band with my script, i just use
1. Open the band with `rioxarray.open_rasterio()`
2. Divides the value per 10000 for reflectance
3. Convert it into `np.float32` to have the same format as eoreader
4. Use `rio.reproject` with a `Resampling.bilinear` to have a 10-meters resolution.

I probably miss one step to achieve your results.

**Reply by remi-braun** [reply]
Yes you are right, maybe this is an issue 😅 
However, collocate only deals with very small interpolations and nearest is faster.
(I have to do it because xarray doesn't round coordinates and therefore attributes NaNs if they are slightly different)

**Reply by remi-braun** [reply]
I see an issue.
What is the processing baseline of your product ? If it is > 04.00, you have to deal with an offset !
See [here](https://sentinel.esa.int/web/sentinel/technical-guides/sentinel-2-msi/processing-baseline) at processing baseline 04.00.

The baseline can be spotted in the name of the .SAFE folder: 
![2023-07-27_15h45_38](https://github.com/sertit/eoreader/assets/67311115/9558f5fe-5d1c-4f0a-9c70-57f7e10e84b1)

**Reply by bcerripromethee** [reply]
I already thought about that issue but i am not sure but Element84 already corrects the offset. In addition, the NIR band is not affected by the issue. It only appears for SWIR1 & SWIR2 bands (so bands which need a resampling process). May be it's because you round the values, for example, a pixel could have :
- Element84 : 0.155506
- Eoreader : 0.1555

But DataArrays in np.float32 "doesn't like" to be rounded with round method (https://stackoverflow.com/a/75090276)

**Reply by remi-braun** [reply]
Ho this is an issue... I really don't like this float64 thing. It shouldn't need float64 arrays when converting uint16 to float...
Could you quantify the errors between your method and EOReader's ?

Note: To reproduce EOReader method, I would reproject to 10m before divinding by 10k and converting to float32

**Reply by bcerripromethee** [reply]
Apparently, by `reprojecting -> dividing -> converting` (in this order), it solves the issue ! It is a little bit mysterious for me but thank you very much !!!!

**Reply by remi-braun** [reply]
Ok, let's stop here then, but I really don't like this mystery haha 😅

**Reply by bcerripromethee** [reply]
Yeah i totally understand !

I reproduced the bug with my old method (odered specified her : https://github.com/sertit/eoreader/discussions/103#discussioncomment-6594417)

There is the values, i had for random pixel in the image : 
**SWIR_2**
_Pixel 1 :_ 
- Element 84 : 0.263994
- Eoreader : 0.264
- Difference of : **0,000006**

_Pixel 2 :_
- Element 84 : 0.1484
- Eoreader : 0.148444
- Difference of : **-0,000044**

_Pixel 3 :_
- Element 84 : 0.337419
- Eoreader : 0.3374
- Difference of : **0,000019**

It's like a random error, but i think it's not ! It musts be related to the format number of the Element84 pixels.

If you want, i can send you the references image + a ratio image of their difference

**Reply by remi-braun** [reply]
No no need to give me the samples.
In any case, values after 10^-5 shouldn't be taken into account. Do you think I should implement a 10^-5 rounding after any reprojection?

**Reply by bcerripromethee** [reply]
Honestly, this is a quite difficult question to answer. Perhaps a good idea would be to add a parameter with a default value (=your current rounding). With this parameter, it will be easier for users to set it as they wish. No ?

**Reply by remi-braun** [reply]
The issue with this is I don't want to complicate EOReader too much (and to many parameters are hard to test).
I will leave it as it is, and wait for other people that may need similar parameters before acting (if it's fine with you)

**Reply by bcerripromethee** [reply]
Yeah, no problem with it, i totally understand !
When I see the Discussion topic, I can see that I spammed you a little bit, sorry for this :sweat_smile:

**Reply by remi-braun** [reply, marked as answer]
In fact, I like those conversations because it always offers another perspective to my work and to other usecases 😉 
I created an issue related to this rounding topic, don't hesitate to write a comment in it. (#105)
I won't do anything about that right now, even if I think this should be done (the reprojection shouldn't have values with another precision than the initial one!)

**Comment by remi-braun** [comment]
In EOReader 0.21.10, some flexibility has been added to resampling:
- Possibility to override the default resampling by setting `"EOREADER_BAND_RESAMPLING"` env variable
```python
os.environ["EOREADER_BAND_RESAMPLING"] = str(Resampling.nearest)
```
- `resampling` argument has been freed in `load` and `stack` method
```python
prod.load(RED, window=window_path, pixel_size=60, resampling=Resampling.nearest)
```

You can see the test function here: https://github.com/sertit/eoreader/blob/ce6ad4bddb0039720a1b36301f1d93c30f104ea8/ci/on_push/test_others.py#L427
